docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,730 @@
+* **Permissions**: caller must be an administrator.
+* **Enterprise requirement**: valid license required.
+
+#### Grant/revoke privileges to/from a role
+
+The `GRANT` and `REVOKE` statements used to grant privileges to a user will accept a role. No syntax
+changes are necessary.
+
+An enterprise license is not required.
+
+#### GRANT role
+
+```
+GRANT rolename TO name [ WITH ADMIN OPTION ]
+```
+
+* **Behavior**: adds `name` as a member of `rolename`. Member is a role admin if `WITH ADMIN OPTION` is specified.
+Fails if either `rolename` or `name` does not exist, if the membership exists, or if this creates a
+membership loop.
+* **Permissions**: must be an administrator or role admin.
+* **Enterprise requirement**: valid license required.
+
+See [Adding and removing admin setting](#adding-and-removing-admin-setting) for behavior details.
+
+#### REVOKE role
+
+```
+REVOKE [ ADMIN OPTION FOR ] rolename FROM name
+```
+
+* **Behavior**: removes `name` as a member of `rolename`. Member admin setting is removed if `ADMIN OPTION FOR` is specified.
+Fails if either `rolename` or `name` does not exist, or if the membership does not exist.
+* **Permissions**: must be an administrator or role admin.
+* **Enterprise requirement**: valid license required.
+
+See [Adding and removing admin setting](#adding-and-removing-admin-setting) for behavior details.
+
+#### SHOW ROLES
+
+```
+SHOW ROLES
+```
+
+* **Behavior**: lists all existing roles (role names only).
+* **Permissions**: none required.
+* **Enterprise requirement**: none required.
+
+#### SHOW GRANTS ON ROLE
+
+```
+SHOW GRANTS ON ROLE rolename [ FOR name ]
+```
+
+Where `rolename` can be one or more role, or `*` and `name` can be one of more user or role.
+
+* **Behavior**: lists role memberships for matching roles/names.
+* **Permissions**: none required.
+* **Enterprise requirement**: none required.
+
+See [Listing expanded role memberships](#listing-expanded-role-memberships) for possible variations.
+
+### Determining role memberships
+
+TODO(mberhault): flesh out this section a bit more, including code samples.
+
+Expanding role memberships is necessary to perform the permission checks listed above.
+
+A naive approach (looking up memberships for every permission check) would be prohibitively expensive.
+
+Instead, we propose the following:
+* per-node cache of already resolved memberships
+* if the user's expanded roles is not cached, look it up
+* the cache holds a lease on the `role_members` table
+* when the `role_members` table descriptor version changes, expire the cached data or refresh it
+* any time a transaction is committed with modification to the `role_members` table: increase the descriptor version
+
+This insures that we can keep role membership information reasonably current, with refreshes being forced
+only when changes occur. Given the small size of membership information, applicable cache expiration will most
+likely be time based (although size maximum should apply).
```

Comment:
Nothing need be time based. All you have to do is always grab a lease on the table descriptor before using the roles, etc. The lease inherently has an expiration time and is refreshed asynchronously. After using the roles, etc, the lease is released. In essence, you have to do nothing here! When you grab a lease and see that the version has changed, the cache can be flushed and ```ModificationTime``` on the cache updated. The cache is then populated lazily

Reply:
Sure, but we have to determine when to stop storing a given a user's expanded roles in the node-level cache. As long as a user is in the cache, we keep refreshing its roles when leases expire.
A size limit may be sufficient as even a few MBs will be a lot of users, but time-based expiration doesn't hurt.